lib_test/bap_project/test_project.ml

```diff
@@ -0,0 +1,92 @@
+open Core_kernel.Std
+open OUnit2
+open Format
+open Word_size
+open Bap.Std
+open Project
+
+let arm_mov_r2_0 = "\x00\x20\xA0\xE3"
+let x86_mov_mem_esp_esi = "\x89\x34\x24"
+
+let str_printer expect_err = Sexp.to_string_hum (sexp_of_string expect_err)
+
+let create_addr = function
+  | W32 -> Addr.of_int ~width:32
+  | W64 -> Addr.of_int ~width:64
+
+let create_mem ?(addr_size=W32) ?(start_addr=0x00)
+    ?(size=0x10) ?(hexstring="\x61\x61\x61\x61") () =
+  let addr = create_addr W32 start_addr in
+  let hexstring = Bigstring.of_string hexstring in
+  let m = Memory.create LittleEndian addr hexstring in
+  match m with
+  | Ok m -> m
+  | Error err -> assert_failure (Error.to_string_hum err)
+
+let create_project architecture =
+  let mem = create_mem () in
+  let dis =
+    disassemble architecture mem in
+  let symtab = Table.add Table.empty mem ".text" |> ok_exn in
+  {arch = architecture; disasm = dis; memory = Memmap.empty;
+   storage = String.Map.empty; symbols = symtab;
+   base = mem }
+
+let substitute project ~kind ~hexstring =
+  let memory =
+    let tag =(Tag.create text kind) in
+    Memmap.add project.memory (create_mem ~hexstring ()) tag in
+  let project_result = Project.substitute {project with memory} in
+  Memmap.fold ~init:[] project_result.memory ~f:(fun acc tag ->
+      match Tag.value text tag with
+      | None -> acc
+      | Some text -> text :: acc)
+
+let test_substitution ctxt ~arch ~kind ~expect ~hexstring =
+  let project = create_project arch in
+  let result = substitute project ~kind ~hexstring in
+  let msg = sprintf "Incorrect %s substitution" kind in
+  assert_equal ~ctxt ~printer:str_printer ~msg expect
+    (match List.hd result with
+     | Some x -> x
+     | None -> assert_failure "Empty result")
+
+let expect_x86_mov_mem_esp_esi =
```

Comment:
or, instead of regular expressions, you can normalize string by removing all whitespaces (including newlines) and compare normalized string. e.g., 
```
{
    mem32 := mem32\
            with [low:32[ESP], el]:u32 <- low:32[ESI]\
}
```
would be transformed to:

```
{mem32:=mem32with[low:32[ESP],el]:u32<-low:32[ESI]}
```